#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv('/content/drive/MyDrive/ANALYTICS/Colaboratory/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby('Branch')['gross income'].sum().round(2)

,gross income
Branch,
A,5057.16
B,5057.03
C,5265.18


Найбільш прибутковим є філіал С - із сумою 5265,18

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [6]:
df.groupby(['City', 'Branch'])['gross income'].sum().round(2)
# я думав, що в кожному місті є різні філіали, тому пробував шукати трохи
# по-іншому, але за підказки АІ залишаю цей варіант, адже в нас чисто
# закріплено, що одне місто - один філіал

,,gross income
City,Branch,
Mandalay,B,5057.03
Naypyitaw,C,5265.18
Yangon,A,5057.16


У місті Naypyitaw, у філіалі С найвищий прибуток - 5265,18. Тому там доречно розмістити ще один магазин.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [7]:
invoices_by_category = df.pivot_table(index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc = 'count')
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [8]:
invoices_by_category[['Electronic accessories']].sort_values(by = 'Electronic accessories', ascending=False)
# AI підказав, що можу скористатись sort_values

Product line,Electronic accessories
Branch,
A,60
B,55
C,55


Бачимо, що найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" в філіалі А - 60 покупок.


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1) У яких філіях частіше купують чоловіки, а в яких жінки?

2) Чи існує гендерна відмінність у виборі категорій товарів?

3) Який спосіб оплати є найпопулярнішим серед клієнтів, і чи відрізняється цей вибір у звичайних покупців та членів клубу?

In [23]:
# У яких філіях частіше купують чоловіки, а в яких жінки?

hypothesis_1 = df.groupby(['Branch', 'Gender'])['Invoice ID'].count()
hypothesis_1

Branch  Gender
A       Female    161
        Male      179
B       Female    162
        Male      170
C       Female    178
        Male      150
Name: Invoice ID, dtype: int64

In [43]:
hypothesis_1_perc = df.groupby('Branch')['Gender'].value_counts(normalize=True).round(4)*100
hypothesis_1_perc

Branch  Gender
A       Male      52.65
        Female    47.35
B       Male      51.20
        Female    48.80
C       Female    54.27
        Male      45.73
Name: proportion, dtype: float64

In [25]:
# Чи існує гендерна відмінність у виборі категорій товарів?

hypothesis_2 = df.pivot_table(index = 'Gender', columns = 'Product line', values = 'Invoice ID', aggfunc = 'count')
hypothesis_2

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Gender,,,,,,
Female,84,96,90,64,79,88
Male,86,82,84,88,81,78


In [41]:
hypothesis_2_perc = df.groupby('Product line')['Gender'].value_counts(normalize=True).round(4)*100
hypothesis_2_perc

Product line            Gender
Electronic accessories  Male      50.59
                        Female    49.41
Fashion accessories     Female    53.93
                        Male      46.07
Food and beverages      Female    51.72
                        Male      48.28
Health and beauty       Male      57.89
                        Female    42.11
Home and lifestyle      Male      50.62
                        Female    49.38
Sports and travel       Female    53.01
                        Male      46.99
Name: proportion, dtype: float64

In [31]:
# Який спосіб оплати є найпопулярнішим серед клієнтів, і чи відрізняється цей вибір у звичайних покупців та членів клубу?

hypothesis_3 = df.pivot_table(index = 'Customer type', columns = 'Payment', values = 'Invoice ID', aggfunc = 'count')
hypothesis_3

Payment,Cash,Credit card,Ewallet
Customer type,,,
Member,168,172,161
Normal,176,139,184


In [35]:
hypothesis_3_perc = df.groupby('Customer type')['Payment'].value_counts(normalize=True).round(4)*100
hypothesis_3_perc

Customer type  Payment    
Member         Credit card    34.33
               Cash           33.53
               Ewallet        32.14
Normal         Ewallet        36.87
               Cash           35.27
               Credit card    27.86
Name: proportion, dtype: float64

1) Дасть змогу зрозуміти, в яких філіях які пропозиції краще просувати — чоловічі/жіночі. Для прикладу, у філії C більше жінок — відповідно більше жіночих пропозицій. У А і В — чоловіків, відповідно більше чоловічих пропозицій.

2) Можна зрозуміти, які категорії товарів краще рухаються по гендерам і які пропозиції краще створювати. Для прикладу, Модні аксесуари, а також Їжа та напої більше для жінок, адже їх частка більша. Для чоловіків більше Здоров'я/краса (де їх частка аж 57.89%), що, до речі, навіть є трохи дивним і цікавим моментом. Можна на цьому зробити акцент, ніби краса та догляд не мають гендеру — створювати пропозиції для чоловіків (перелік — такі/такі і тд).

3) Побачити спосіб розрахунку клієнтів. Для прикладу, члени клубу дають перевагу кредитним карткам. Можна створити наступну пропозицію — за розрахунок карткою Мастеркард 5% кешбек або щось подібне. У звичайних покупців перевага у електронних гаманцях — щось подібне або навпаки промотивувати тим, що стаєте членом клубу - отримуєте якийсь бонус до конкретного способу оплати і тд.